Error Propagation

lciorciaro opened the pull request:
This resolves #18. 
Error propagation from a broken node to nodes that depend on it is made possible. For this a new parameter `children` is added to the node structure (`dependencies` has been renamed `parents` to be consistent), which contains a list of all nodes that depend on this node. The new `add_parent` and `remove_parent` methods take care of adding and removing references in the `children` lists of other nodes where needed, meaning that the `children` should never have to be edited manually. Additionally, the `close` method is overridden to remove a node from the `children` of its parents when it is deleted.

Commit:
Added error propagation

Includes new tests. Tests for loading graph need to be updated, because
the parameter name changed from dependencies to parents.

Commit:
Updated test graphs

Updated test graphs to use new parameter names 'parents' and 'children'
instead of 'dependencies'. There is a script now to automatically
construct these test graphs and save them to files.

Commit:
Updated example notebook and node methods

Decided that the nodes should only methods for adding and removing
parents, not children.

Commit:
Fixed docstring

Commit:
Fixed codacy issues

AdriaanRol commented on autodepgraph/graph.py:
```diff
@@ -33,7 +33,9 @@ def load_graph(self, filename, load_node_state=False):
 
-            pars_to_update = ['dependencies', 'check_functions',
+            # children are not set because it is automatically set when
+            # parents are set.
```
I would change this comment to make clear that children are set when this node gets added as a parent to another node.

AdriaanRol commented on autodepgraph/node.py:
```diff
@@ -27,8 +28,15 @@ def __init__(self, name, verbose=False):
+                                     + 'on which this node depends',
+                           initial_value=[],
+                           vals=vals.Lists(vals.Strings()),
+                           parameter_class=ManualParameter)
```
I would not make this a manual parameter as it should only be modified using the add/remove parent functions. I would set a get_cmd to return the self._parents and make the docstring refer to the add/remove parent functions. 
Alternatively (and maybe even better) you could support setting a list of strings but make sure that the set_cmd iterates over the elements in the list using the add parent function.

AdriaanRol commented on autodepgraph/node.py:
```diff
@@ -27,8 +28,15 @@ def __init__(self, name, verbose=False):
+                                     + 'which depend on this node',
                            initial_value=[],
                            vals=vals.Lists(vals.Strings()),
                            parameter_class=ManualParameter)
```
I would also make this a getable only command with reference on how to set it in the docstring.

AdriaanRol commented on autodepgraph/node.py:
```diff
@@ -75,8 +86,73 @@ def _get_state(self):
+        '''
+        Removes reference to this node from its parents and close the
+        instrument.
+        Note: References to this node from child nodes are not removed
```
good comment

AdriaanRol commented on autodepgraph/tests/write_test_graphs.py:
```diff
@@ -0,0 +1,57 @@
+import autodepgraph as adg
+import autodepgraph.graph as g
+import autodepgraph.node as n
+import os
```
I guess this corresponds to creating the datafiles that are used in the tests. Maybe add some comment as to what this file is used for.

AdriaanRol commented on autodepgraph/node.py:
```diff
@@ -75,8 +86,73 @@ def _get_state(self):
+        except KeyError:
+            logging.warning('Parent node "{}" not found.'.format(name))
+
+    def propagate_error(self, state):
```
I have not found the test for this yet, will that be there?

AdriaanRol reviewed (changes requested):
See comments.

Commit:
Adressed issues raised by @AdriaanRol

Issues were:
- parents and children are not ManualParameters anymore, children are
  only gettable
- improved some comments for clarity
- added test for error propagation

Additionally, I improved the add_parent and remove_parent method: The
argument can now be a string, a node object, or a list/numpy.array of
these types.

Commit:
Codacy: type check with isinstance

AdriaanRol commented on autodepgraph/graph.py:
```diff
@@ -33,8 +33,10 @@ def load_graph(self, filename, load_node_state=False):
-            # parents are set.
+            # children is not in pars_to_update because the children are set
+            # whenever a node is added as a parent to another node. This means
+            # that loading and setting parents for all node automatically also
```
for all node automatically > for all node-**s**- automatically

typo missing s

AdriaanRol commented on autodepgraph/node.py:
```diff
@@ -30,16 +30,24 @@ def __init__(self, name, verbose=False):
+                                     'Children are automatically added '
+                                     'whenever this node is added as a '
+                                     'parent to another node.',
+                           set_cmd=self._set_children,
```
If you do not specify a set_cmd then it will just raise an error when you try to set it (as you would want) and there is no need to create the empty function below.

AdriaanRol commented on autodepgraph/node.py:
```diff
@@ -86,6 +94,24 @@ def _get_state(self):
+        Sets the parents parameter by calling the add_parent method for every
+        item. This ensures that this node is also added to the children of
+        the other nodes.
+        '''
```
This function does not set the parameter to equal the value (list of strings) you set unless it was an empty list of parents. 
Instead it adds items to the list of parents. This results in the wrong value being contained in the snapshot (and potentially weird artefacts when saving and plotting the graph). 

I would suggest either making this such that the list after setting corresponds to exactly what is set or removing the set_parents functionality (just like I would remove the set children function completely).

AdriaanRol reviewed (changes requested):
Some comments to improve robustness and prevent hard to track bugs in the set parent command.

lciorciaro committed:
Removed set commands for parents and children

lciorciaro committed:
Added set command for parents again

This is needed such that the parents parameter can be treated the
same as other parameters in the load function for the graph.
_set_parents now removes all elements from parents first, such that
in the end parents is exactly equal to the argument given.

lciorciaro committed:
Fixed test

Test was attempting to explicitly set children, which is not allowed.
